tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/dataset/QueryDataSet.java

```diff
@@ -107,17 +112,59 @@ protected void initQueryDataSetFields(
     }
   }
 
+  public Set<Integer> getWithoutNullColumnsIndex() {
+    return withoutNullColumnsIndex;
+  }
+
+  public void setWithoutNullColumnsIndex(Set<Integer> withoutNullColumnsIndex) {
+    this.withoutNullColumnsIndex = withoutNullColumnsIndex;
+  }
+
   public boolean hasNext() throws IOException {
     // proceed to the OFFSET row by skipping rows
     while (rowOffset > 0) {
       if (hasNextWithoutConstraint()) {
         RowRecord rowRecord = nextWithoutConstraint(); // DO NOT use next()
-        // filter rows whose columns are null according to the rule
-        if ((withoutAllNull && rowRecord.isAllNull())
-            || (withoutAnyNull && rowRecord.hasNullField())) {
-          continue;
+        if (withoutNullColumnsIndex.isEmpty()) {
+          // filter rows whose columns are null according to the rule
+          if ((withoutAllNull && rowRecord.isAllNull())
+              || (withoutAnyNull && rowRecord.hasNullField())) {
+            continue;
+          }
+          rowOffset--;
```

Comment:
extract this commone statement in `if` and `else` block.

Reply:
fix